File: 2024/day-01/day-01.fs
module Day01

open Utils
open FSharpx
open FSharpx.Text

let parse (input : string list) : int64 list * int64 list =
    let l =
        List.map (Strings.split ' ' >> Seq.choose FSharpOption.ParseInt64 >> Seq.assertPairs) input

    (List.map fst l, List.map snd l)

let solvePart0 (input : string list) : int64 =
    let left, right = parse input
    let zips = List.zip (List.sort left) (List.sort right)
    List.sum (List.map (fun (x, y) -> abs (x - y)) zips)

let solvePart1 (input : string list) : int64 =
    let left, right = parse input

    let occurrences =
        List.fold (fun acc k -> Map.change k (fun v -> Some (defaultArg v 0L + 1L)) acc) Map.empty right

    let scores = List.map (fun x -> x * defaultArg (Map.tryFind x occurrences) 0L) left
    List.sum scores

type ThisDay() =
    interface IDay with
        member this.day () =
            Day.create 01 solvePart0 solvePart1
            <| seq {
                "input-ex0.txt", Some (11, 31)
                "input-real0.txt", Some (765748, 27732508)
            }

File: 2015/day-01/day-01.c
#include "common/common.h"

static int64_t solvePart0(FILE *f) {
    int64_t floor = 0;
    int c;
    while ((c = fgetc(f)) != EOF) {
        if (c == '(') {
            floor++;
        } else if (c == ')') {
            floor--;
        }
    }
    return floor;
}

static int64_t solvePart1(FILE *f) {
    int64_t floor = 0;
    int64_t pos = 0;
    int c;
    while ((c = fgetc(f)) != EOF) {
        if (c == '(') {
            floor++;
        } else if (c == ')') {
            floor--;
        }
        pos++;
        if (floor < 0) {
            return pos;
        }
    }
    return -1;
}

static int dayMain() {
    int failed = 0;
    failed += checkInputInt("day-01/input-real0.txt", solvePart0, 280, solvePart1, 1797);
    return failed;
}

daySolver day01 = {1, dayMain, true};
